### PR TITLE
Add device asserts when mask == None

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100880

See https://github.com/pytorch/pytorch/pull/100816#issuecomment-1537515672

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1082,10 +1080,12 @@ def print_min(expr):
             # It'd be less # error-prone to use and/or/not, which is suported by triton
             cond = f"((0 <= {var}) & ({var} < {size}))"
             cond_print = f"0 <= {var} < {size}"
-            if not isinstance(original_index, sympy.Integer):
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1085]: This condition was unnecessary, as it's never going to be true when we have indirect indexing.

**Comment by lezcano** [comment]
The errors would be fixed by https://github.com/openai/triton/pull/1641. Now, I think I'll better add this fix to https://github.com/pytorch/pytorch/pull/100550, as the new `device_assert`s that it would create on static values would mostly be removed by the value analysis in that PR.

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1082,10 +1080,12 @@ def print_min(expr):
...
+            if mask != "None" and not isinstance(original_index, sympy.Integer):
                 var_mask = f"({mask})" if "&" in mask else mask
                 var_mask = f" | ~{var_mask}"
             else:
+                # mask may be None if, for example, xnumel == 1 or xnumel % 2048 == 0
```
Comment (ngimel) [inline torch/_inductor/codegen/triton.py:1087]: are we too trigger-happy with device asserts now? As @Jokeren points out, in many cases with `original_index` known at compile time (which by far is the most common case) the mask will be None (it's a single value, we specifically remove mask for this case) and we will know at compile time whether it's valid or not (we are checking that during lowering/decomp, and invalid cases should be weeded out there, I'd say only valid values should ever make it to codegen).

**Comment by ngimel** [comment]
The case I was concerned about didn't involve device asserts on scalar values. I think we should be doing value analysis and erroring out as soon as possible, we shouldn't make it to codegen with still invalid statically known values. 
In fact, can I suggest an opposite direction - we should be generating device asserts only for data that we legit read from inputs and thus have no idea what it will be. Indices that we compute (and thus can simplify with value analysis) should never have gotten device asserts in the first place, what can be done before codegen should be done before codegen.

**Pull request closed without merging**